plugins/modules/postgresql_user.py

```diff
@@ -920,6 +957,71 @@ def add_comment(cursor, user, comment, check_mode):
         return False
 
 
+def compare_user_configurations(current, desired):
+    """Compares two configurations and returns a list of values to reset as well as a dict of parameters to update."""
+    reset = []
+    update = desired.copy()
+
+    # check each item in the current configuration
+    for key, value in current.items():
+        # we already have the correct setting
+        if key in desired and value == desired[key]:
+            # so we can remove it from the list
+            del update[key]
+        # if the key is not in the list of settings we want
+        elif key not in desired:
+            # we will reset it on the database
+            reset.append(key)
+        # if the setting is not in the db or has the wrong value, it will get updated
+
+    return {"reset": reset, "update": update}
+
+
+def parse_user_configuration(module, configs):
+    """Parses configuration from a list of 'key=value' strings like returned from the database to a dict."""
+    if configs is not None:
+        try:
+            # parses a list of "key=value" strings to a dict
+            return {t[0]: t[1] for t in map(lambda s: s.split("=", 1), configs)}
+        except IndexError:
+            module.fail_json(
+                msg="Expecting a list of strings where each string has the format 'key=value'.")
+    else:
+        return {}
+
+
+def user_configuration(cursor, module, user, configuration):
+    """Updates the user's configuration parameters if necessary."""
+    current_config_query = "SELECT rolconfig FROM pg_roles WHERE rolname = %(user)s;"
+    cursor.execute(current_config_query, {"user": user})
+    current_config = cursor.fetchone()
+    changed = False
+
+    if current_config is None:
+        module.fail_json(msg="Can't find user %(user)s in 'pg_roles'" % {"user": user})
+
+    current_config_dict = parse_user_configuration(module, current_config['rolconfig'])
+    config_updates = compare_user_configurations(current_config_dict, configuration)
+
+    try:
+        # It seems psycopg's prepared statements don't work with 'ALTER ROLE' at this point.
+        # This is vulnerable to SQL-injections (added to docs) but I don't see a better way to do this.
```

Comment:
psycopg is serializing values as _constants_ (numbers, `'`-quoted strings, ...), but `alter role` expects _identifiers_ (`"`-quoted).
This project already dealt with SQL-injection prevention for identifiers, please use the existing [`pg_quote_identifier`](https://github.com/ansible-collections/community.postgresql/blob/ea28b3de22464d419d299722dd250033798886c6/plugins/module_utils/database.py#L125) function for identifiers (role name, config variable name).

psycopg should serialize the config variable value.

Thereby, this can be made safe against SQL injections.

Reply:
I tried, but [`pg_quote_identifier`](https://github.com/ansible-collections/community.postgresql/blob/ea28b3de22464d419d299722dd250033798886c6/plugins/module_utils/database.py#L125) only allows one dot, while role names, as well as configuration parameters may include more than one dot.  
That is probably also why it is not used in [`user_add`](https://github.com/ansible-collections/community.postgresql/blob/ea28b3de22464d419d299722dd250033798886c6/plugins/modules/postgresql_user.py#L386), for example.

Reply:
Well ok, `pg_quote_identifier` could be better documented. It actually supports both unquoted and already-quoted inputs as well as chained(?) identifiers like `database.schema.table.column`. It has a parameter for how many identifier parts are allowed.

In this case we could also only support unquoted input and escape it like this:
```python3
sql_safe = '"%s"' % (unsafe_input.replace('"', '""'), )
```
As https://www.postgresql.org/docs/current/sql-syntax-lexical.html explains, this should be enough to safely write an identifier.

Reply:
There are several issues:  
- One is in [this `dict`](https://github.com/ansible-collections/community.postgresql/blob/ea28b3de22464d419d299722dd250033798886c6/plugins/module_utils/database.py#L43). It specifies that a `role` may only have one dot and there is no entry for `parameter` or something like that.
- Another one is, that PostgreSQL doesn't recognize a user if quoted in the way `pg_quote_identifier` returns:
  ```
  postgres=# GRANT CONNECT ON DATABASE "postgres" TO "test"."user" ;
  ERROR:  syntax error at or near "."
  LINE 1: GRANT CONNECT ON DATABASE "postgres" TO "test"."user" ;
                                                        ^
  postgres=# GRANT CONNECT ON DATABASE "postgres" TO "test.user" ;
  GRANT
  ```

Also this would still be vulnerable to SQL-injections in the value even if we quoted the user and parameter name.

I'm using `check_input` when `trust_input` is `false`. It couldn't properly check dicts, so I changed it to be able to check lists, tuples and dicts recursively.  
Imho it is good enough to run the input through `check_input` when `trust_input=false`.

Reply:
`_PG_IDENTIFIER_TO_DOT_LEVEL` specifies how many dot-separated parts are allowed, which is one more than how many dots are allowed. For role names, no dots are allowed. Role names are single identifiers, nothing schema-qualified or whatnot.

I think the `check_input` function doesn't help much because it only checks for some specific SQL injections. This is no guarantee to be injection-safe and gives a false sense of trust.

The only way to get protected against SQL injections is by proper input escaping. This is very simple: just double every occurrence of `"`, just as I suggested in the comment earlier.

`pg_quote_identifier` adds some magic around that simple escaping to detect already-quoted input.
If you disklike that magic, ok, don't use it.
But please implement proper input escaping.

Reply:
> For role names, no dots are allowed.

That is not correct. Dots are allowed, but they don't have the same meaning as, for example, in a table-name, where you have `database.schema.table`. So a username `some.user.name` is a valid username, but you can't quote it like this `"some"."user"."name"`. This is what `pg_quote_identifier` does and it simply doesn't work. This has nothing to do with me wanting or not wanting to use that function.

> The only way to get protected against SQL injections is by proper input escaping. This is very simple: just double every occurrence of ", just as I suggested in the comment earlier.

Sorry, but I consider this a very dangerous belief. Relying on escaping is exactly what made, for example, unicode homoglyph sql-injections possible (I assume it is not possible here, but it is an example where trusting in escaping has been tripping people before). Escaping is imho not fit to reliably prevent SQL injections and especially writing custom escaping-functions is prone to oversight and is what actually gives you the false sense of security.  
The "better" (and imho only) way to prevent SQL injections is prepared statements (which doesn't work in this case).
Here is, by the way, what happens when you would do the double-`"` thing in psql (I find this funny):
```
postgres=# create user some""user; select 1;--";
ERROR:  syntax error at or near "some"
LINE 1: create user some""user;
                    ^
 ?column?
----------
        1
(1 row)
```
The first statement throws an error and the second one still gets executed. I wasn't able to reproduce this with psycopg, though.


In any case, you are the reviewer, so if you want it done the way you write, I have no problem doing it that way.  
There is, though, the issue, that this quoting isn't done anywhere else in this module (for example [here](https://github.com/ansible-collections/community.postgresql/blob/ea28b3de22464d419d299722dd250033798886c6/plugins/modules/postgresql_user.py#L386)).  
This leads to the following problems:
If I would only add that to my code, other points (like the one in `user_add`) would still be vulnerable.
I could add this quoting in all places throughout the module (Idk if that is in the scope of this MR), but in that case, we would break backwards-compatibility.  
Somebody who already uses the double-`"` method to create users with quotes might be using `user: 'some""user'` to create a user `some"user`. But if I now start doing the replace-thing, this would lead to the user being created turn into `some""user`, as the input `'some""user'` would get turned into `some""""user`.  
If I only add the replacing to my code, the module will break, as in the existing code, the user would be `some"user` but in my code, it would be `some""user`.  
I could use a regex to only replace single instances of `"` but in that case, we will be vulnerable to SQL injections again, if somebody does `user: 'someuser"""; drop database postgres;--'`.

Bottom line:  
I don't see a consistent way to do this without breaking backwards-compatibility.

My opinion:  
This kind of escaping hasn't been done until now, and if the plan is to do it, it should be done properly, as in a full refactoring where it is done throughout the whole module, but that would probably need to be done in a major-release as it will backwards-compatibility as shown above.

Reply:
> For role names, no dots are allowed.

Yes, PostgreSQL allows dots in role names. What I wanted to say: For role names, `pg_quote_identifier` only allows a single identifiers, not multiple dot-separated identifiers.
For example:
```python3
>>> pg_quote_identifier('a', 'role')
'"a"'
>>> pg_quote_identifier('"a"', 'role')
'"a"'
>>> pg_quote_identifier('a.b', 'role')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "database.py", line 128, in pg_quote_identifier
    raise SQLParseError('PostgreSQL does not support %s with more than %i dots' % (id_type, _PG_IDENTIFIER_TO_DOT_LEVEL[id_type]))
quote.SQLParseError: PostgreSQL does not support role with more than 1 dots
>>> pg_quote_identifier('"a.b"', 'role')
'"a.b"'
>>> pg_quote_identifier('a"', 'role')
'"a"""'
>>> pg_quote_identifier('💣', 'role')
'"💣"'
```

Reply:
The example `create user some""user; select 1;--";` doesn't apply because proper escaping would add surrounding quotes in addition to doubling quotes from the input:
```sql
postgres=# create user "some""user; select 1;--";
CREATE ROLE
postgres=# select rolname from pg_roles where rolname like 'some%';
        rolname         
------------------------
 some"user; select 1;--
```

Reply:
> This kind of escaping hasn't been done until now, and if the plan is to do it, it should be done properly, as in a full refactoring where it is done throughout the whole module, but that would probably need to be done in a major-release as it will backwards-compatibility as shown above.

Escaping by doubling quotes has been done [here](https://github.com/ansible-collections/community.postgresql/blob/ea28b3de22464d419d299722dd250033798886c6/plugins/modules/postgresql_privs.py#L797) and [here](https://github.com/ansible-collections/community.postgresql/blob/ea28b3de22464d419d299722dd250033798886c6/plugins/module_utils/database.py#L106). The latter is part of `pg_quote_identifier`.

Yes, usage is not so consistent throughout the collection and handling all identifier input equal would be ideal.
Yes, making it consistent would require a major release.
I would like that but I don't think this is urgent. -> separate issue

For this PR I would like to see some input escaping instead of none. Preferably with one of the [two methods I described earlier](https://github.com/ansible-collections/community.postgresql/pull/722#discussion_r1688636681).

Reply:
Okay, now I understand, that the role-name needs to be "pre-quoted" before `pg_quote_identifiers` accepts it. I had to change other parts of the code as well to make sure the results are consistent, but I think I managed to get it to work. (see bc873bb)  
I didn't touch the (deprecated) privileges part in the users-module, though, so that is still vulnerable.

I'm also sanitizing the keys and values which should prevent most sql-injections. (see 540a89b)